### PR TITLE
SIMSBIOHUB-445: Feature Branch - Consolidate Observations/Telemetry/Animals

### PR DESCRIPTION
**Pull request by curtisupshall**
## Links to Jira Tickets

 - https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-445

## Description of Changes

- Combined the Telemetry and Observations map/table view on the Survey page, allowing users to toggle between them;
- Added server-side support for paginating the observations fetching endpoint;
- Modified the Manage Observations page to use server-side pagination;
- Added a new endpoint to fetch all the spatial data relating to survey observations;
- Deleted some old code related to Survey Summary Results.

## Testing Notes

You should be able to:
- [ ] View observation and telemetry points on the survey page
- [ ] On the survey page, sort on various columns and page through the observations table (except for Species; see [465](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-465))
- [ ] On the Manage Observations page, sort on various columns and page through the observations

### CODE REVIEW

File: app/src/contexts/observationsTableContext.tsx [hunk 1]
```diff
@@ -179,6 +212,42 @@ export const ObservationsTableContextProvider = (props: PropsWithChildren<Record
   // Stores the current validation state of the table
   const [validationModel, setValidationModel] = useState<ObservationTableValidationModel>({});
 
+  // Pagination States
+
+  // const [totalRows, setTotalRows] = useState<number>(0);
```
Comment (curtisupshall) [inline app/src/contexts/observationsTableContext.tsx]: Nit: remove

File: app/src/contexts/observationsTableContext.tsx [hunk 2]
```diff
@@ -179,6 +212,42 @@ export const ObservationsTableContextProvider = (props: PropsWithChildren<Record
...
+      page: paginationModel.page + 1, // +1 to correct an off by one error with pagination
+      limit: paginationModel.pageSize
+    });
+  }, []);
+
+  // Fetch new rows based on sort/ pagination model changes
+  useEffect(() => {
+    const sort = firstOrNull(sortModel);
+    observationsContext.observationsDataLoader.refresh({
+      page: paginationModel.page + 1, // +1 to correct an off by one error with pagination
```
Comment (curtisupshall) [inline app/src/contexts/observationsTableContext.tsx]: Might rephrase this comment:
```javascript
// API pagination pages begin at 1, but MUI DataGrid pagination begins at 0.
page: paginationModel.page + 1,
```

File: app/src/features/surveys/view/SurveySpatialData.tsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  const biohubApi = useBiohubApi();
+  const { projectId, surveyId } = useContext(SurveyContext);
+
+  const observationsGeometryDataLoader = useDataLoader(() =>
+    biohubApi.observation.getObservationsGeometry(projectId, surveyId)
+  );
+
+  observationsGeometryDataLoader.load();
+
+  //TODO: look into adding this to the query param
```
Comment (curtisupshall) [inline app/src/features/surveys/view/SurveySpatialData.tsx]: Per discussion, persisting this across refreshes/page changes is not needed (for now)

File: app/src/features/surveys/view/SurveySpatialData.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+          properties: {},
+          geometry: observation.geojson
+        },
+        popupComponent: undefined
+      };
+    });
+  }, [observationsGeometryDataLoader.data]);
+
+  // TODO: this needs to be saved between page visits
+  // const [layout, setLayout] = useState<SurveySpatialDataLayout>(SurveySpatialDataLayout.MAP);
```
Comment (curtisupshall) [inline app/src/features/surveys/view/SurveySpatialData.tsx]: Remove layout code?

Comment (curtisupshall) [reply app/src/features/surveys/view/SurveySpatialData.tsx]: Removed.

**Review by KjartanE — COMMENTED** [review]
image
Drop down buttons are not functional

**Review by KjartanE — COMMENTED** [review]
image
Split and table view provide no information

**Review by KjartanE — COMMENTED** [review]
image
Survey boundary and sample sites are not present in any view

**Review by jeznorth — DISMISSED** [review]

File: app/src/interfaces/useObservationApi.interface.ts [hunk 5]
```diff
@@ -1,6 +1,22 @@
-import { IObservationRecord } from 'contexts/observationsTableContext';
+import { IObservationRecordWithSamplingData, ISupplementaryObservationData } from 'contexts/observationsTableContext';
 
 export interface IGetSurveyObservationsResponse {
-  surveyObservations: IObservationRecord[];
-  supplementaryObservationData: { observationCount: number };
+  surveyObservations: IObservationRecordWithSamplingData[];
+  supplementaryObservationData: ISupplementaryObservationData;
+  pagination: {
+    total: number;
```
Comment (MacDeluca) [inline app/src/interfaces/useObservationApi.interface.ts:7]: Is this total pages?

File: app/src/interfaces/useObservationApi.interface.ts [hunk 6]
```diff
@@ -1,6 +1,22 @@
...
+    current_page: number;
+    last_page: number;
+    per_page?: number;
```
Comment (MacDeluca) [inline app/src/interfaces/useObservationApi.interface.ts:10]: Is this rows per_page?

File: api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/delete.ts [hunk 7]
```diff
@@ -115,12 +115,9 @@ export function deleteSurveySampleSiteRecords(): RequestHandler {
...
 
-      const observationCount = await observationService.getObservationsCountBySampleSiteIds(
-        surveyId,
-        surveySampleSiteIds
-      );
+      const response = await observationService.getObservationsCountBySampleSiteIds(surveyId, surveySampleSiteIds);
 
-      if (observationCount.observationCount > 0) {
+      if (response.observationCount > 0) {
         throw new HTTP500(`Cannot delete a sample sites that is associated with an observation`);
```
Comment (MacDeluca) [inline api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/delete.ts]: Nit: grammar

**Comment by NickPhura** [comment]
Found one funny interaction, which could probably be fixed in a later PR: If you add a new record on the observation page, but then paginate to a new page, the UI gets into a funny state where the new row is gone, but the save/discard buttons are still active. Similar type issue with the validation banner remaining even though the new row is gone.

**Comment by curtisupshall** [comment]
> Found one funny interaction, which could probably be fixed in a later PR: If you add a new record on the observation page, but then paginate to a new page, the UI gets into a funny state where the new row is gone, but the save/discard buttons are still active. Similar type issue with the validation banner remaining even though the new row is gone.

Yeah that's an issue. We might not need to solve this if we add rows through a modal - or could solve it with a sticky row, if that ever becomes possible. Or we could even disable pagination entirely while the table is in "edit" mode.

**Review by NickPhura — DISMISSED** [review]
Played with it a bit locally (not super exhaustively tbh). And gave the code a quick look over. Seems to function well. The endpoints look tightly defined which is great.

Not specifically about this PR, but in general:
One thing we should get into the habit of including in our OpenApi spec, is adding `additionalProperties: false`. Gets added as a sibling to the `properties` field in a `type: object` section. Just helps ensure over time that the openapi stays in sync with the response by making it more strict.

File: app/src/interfaces/useObservationApi.interface.ts (hunk 5 above)
Comment (curtisupshall) [reply app/src/interfaces/useObservationApi.interface.ts:7]: `total` is the total number of rows for the survey. I'll add a comment.

File: app/src/interfaces/useObservationApi.interface.ts (hunk 6 above)
Comment (curtisupshall) [reply app/src/interfaces/useObservationApi.interface.ts:10]: That is correct

File: api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/delete.ts (hunk 7 above)
Comment (curtisupshall) [reply api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/delete.ts]: Good catch

**Review by MacDeluca — APPROVED** [review]
😻

**Review by NickPhura — APPROVED** [review]

**Pull request merged**